From 9298916776187d5071111daee8c82690d7bd91c8 Mon Sep 17 00:00:00 2001
From: Pavel Raiskup <praiskup@redhat.com>
Date: Thu, 9 Feb 2012 20:01:04 +0100
Subject: [PATCH] Fix for long name corruption when --posix used.

Test case included in patch done by Kamil Dudka.

Related bug:    #656834
Upstream discussion:
    http://thread.gmane.org/gmane.comp.gnu.tar.bugs/4152/focus=4160
---
 src/sparse.c       |    7 +++++--
 tests/Makefile.am  |    1 +
 tests/sparse04.at  |   47 +++++++++++++++++++++++++++++++++++++++++++++++
 tests/testsuite.at |    1 +
 4 files changed, 54 insertions(+), 2 deletions(-)
 create mode 100644 tests/sparse04.at

diff --git a/src/sparse.c b/src/sparse.c
index 9fde508..c9d6473 100644
--- a/src/sparse.c
+++ b/src/sparse.c
@@ -1025,8 +1025,11 @@ pax_dump_header_1 (struct tar_sparse_file *file)
   xheader_store ("GNU.sparse.name", file->stat_info, NULL);
   xheader_store ("GNU.sparse.realsize", file->stat_info, NULL);
   
-  file->stat_info->file_name = xheader_format_name (file->stat_info,
-					    "%d/GNUSparseFile.%p/%f", 0);
+  file->stat_info->file_name =
+    xheader_format_name (file->stat_info, "%d/GNUSparseFile.%p/%f", 0);
+  /* Make sure the created header name is shorter than NAME_FIELD_SIZE: */
+  if (strlen (file->stat_info->file_name) > NAME_FIELD_SIZE)
+    file->stat_info->file_name[NAME_FIELD_SIZE] = 0;
 
   blk = start_header (file->stat_info);
   /* Store the effective (shrunken) file size */
diff --git a/tests/Makefile.am b/tests/Makefile.am
index e9b753c..11b8fda 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -126,6 +126,7 @@ TESTSUITE_AT = \
  sparse01.at\
  sparse02.at\
  sparse03.at\
+ sparse04.at\
  sparsemv.at\
  sparsemvp.at\
  spmvp00.at\
diff --git a/tests/sparse04.at b/tests/sparse04.at
new file mode 100644
index 0000000..4d41be7
--- /dev/null
+++ b/tests/sparse04.at
@@ -0,0 +1,47 @@
+# Process this file with autom4te to create testsuite. -*- Autotest -*-
+#
+# Test suite for GNU tar.
+# Copyright (C) 2010 Free Software Foundation, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.
+
+AT_SETUP([storing long sparse file names])
+AT_KEYWORDS([sparse sparse04])
+
+# Description: Tar versions from 1.15.92 to 1.25 would incorrectly
+# store sparse file names longer than 100 characters in pax mode.
+# Namely, the `path' keyword of the produced PAX header would contain the
+# crafted name of the header itself, instead of that of the file.
+# Reported by: Kamil Dudka <kdudka@redhat.com>
+# References: <201011250026.44908.kdudka@redhat.com>,
+#             http://lists.gnu.org/archive/html/bug-tar/2010-11/msg00099.html
+
+m4_define([NAME_111],
+	  [123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960])
+
+AT_TAR_CHECK([
+genfile --sparse --file NAME_111 --block-size 512 8M A || AT_SKIP_TEST
+tar -c --sparse --posix NAME_111 | tar t
+],
+[0],
+[NAME_111
+],
+[],
+[],
+[],
+[pax])
+
+AT_CLEANUP
diff --git a/tests/testsuite.at b/tests/testsuite.at
index f581071..4d9c3fd 100644
--- a/tests/testsuite.at
+++ b/tests/testsuite.at
@@ -205,6 +205,7 @@ m4_include([shortrec.at])
 m4_include([sparse01.at])
 m4_include([sparse02.at])
 m4_include([sparse03.at])
+m4_include([sparse04.at])
 m4_include([sparsemv.at])
 m4_include([spmvp00.at])
 m4_include([spmvp01.at])
-- 
1.7.4.4

